src/ClientImpl.cpp

```diff
@@ -306,37 +354,109 @@ ClientImpl::~ClientImpl() {
     }
 
     event_base_free(m_base);
+
     if (m_wakeupPipe[1] != -1) {
        ::close(m_wakeupPipe[0]);
        ::close(m_wakeupPipe[1]);
     }
+
+    if (m_enableSSL) {
+        if (m_clientSslCtx != NULL) {
+            // free clients ssl context
+            SSL_CTX_free(m_clientSslCtx);
+            m_clientSslCtx = NULL;
+        }
+        cleanupErrorStrings = true;
+    }
+
+    {
+        boost::mutex::scoped_lock lock(m_globalResourceLock);
+        if (--m_numberOfClients == 0) {
+            if (cleanupEvp) {
+                // release global resource allocated for generating digest/hash
+                EVP_cleanup();
+            }
+            if (m_enableSSL) {
+                // unload ssl n crypto error strings
+                ERR_free_strings();
+            }
+        }
+    }
 }
 
 // Initialization for the library that only gets called once
 pthread_once_t once_initLibevent = PTHREAD_ONCE_INIT;
 void initLibevent() {
-    // try to run threadsafe libevent
-    if (evthread_use_pthreads()) {
-        throw voltdb::LibEventException();
+    int status = evthread_use_pthreads();
+    if (status) {
+        std::ostringstream msg;
+        msg << "Failed to initialize libevent in thread-safe mode. Error: " << status;
+        throw LibEventException(msg.str());
     }
 }
 
 const int64_t ClientImpl::VOLT_NOTIFICATION_MAGIC_NUMBER(9223372036854775806);
 const std::string ClientImpl::SERVICE("database");
 
-ClientImpl::ClientImpl(ClientConfig config) throw(voltdb::Exception, voltdb::LibEventException) :
+class CleanupEVP_MD_Ctx{
```

Comment:
In case if the EVP context has any error from the openssl, the error condition is treated fatal and  execution halts by throwing exception to user from the hash function and propagating it to the user. So this seems more simpler to clean acquired resource at different exists points in hash function itself.